src/Proto3/Suite/Class.hs

```diff
@@ -485,51 +484,61 @@ instance MessageField BL.ByteString
+
   protoType _ = messageField (Map (primType (proxy# :: Proxy# k)) (primType (proxy# :: Proxy# v))) Nothing
 
 instance {-# OVERLAPS #-} (Ord k, Primitive k, Named v, Message v, MessageField k) => MessageField (M.Map k (Nested v)) where
-  encodeMessageField num = foldMap (Encode.embedded num . encodeMessage (fieldNumber 1)) . M.toList
+  encodeMessageField num =
+    let encodeField :: k -> Nested v -> Encode.MessageBuilder
+        encodeField k v = Encode.embedded num (encodeMessage 1 (k, v))
+     in M.foldrWithKey (\k -> (<>) . encodeField k) mempty
 
   -- Data.Map.fromList will retain the last key/value mapping. From the spec:
   --
   -- > When parsing from the wire or when merging, if there are duplicate map
   -- > keys the last key seen is used.
-  decodeMessageField = M.fromList . Foldable.toList
-                       <$> repeated (Decode.embedded' (decodeMessage (fieldNumber 1)))
+  decodeMessageField = M.fromList <$> repeated (Decode.embedded' (decodeMessage 1))
+
   protoType _ = messageField (Map (primType (proxy# :: Proxy# k)) (Named . Single $ nameOf (proxy# :: Proxy# v))) Nothing
 
 instance (HasDefault a, Primitive a) => MessageField (ForceEmit a) where
   encodeMessageField = encodePrimitive
 
 instance (Named a, Message a) => MessageField (Nested a) where
-  encodeMessageField num = foldMap (Encode.embedded num . encodeMessage (fieldNumber 1))
-                           . coerce @(Nested a) @(Maybe a)
-  decodeMessageField = coerce @(Parser RawField (Maybe a)) @(Parser RawField (Nested a))
-                       (Decode.embedded (decodeMessage (fieldNumber 1)))
+  encodeMessageField num x =
+    case coerce @(Nested a) @(Maybe a) x of
+      Nothing -> mempty
+      Just x' -> Encode.embedded num (encodeMessage 1 x')
+
+  decodeMessageField = coerce @(Maybe a) @(Nested a) <$> Decode.embedded (decodeMessage 1)
+
   protoType _ = messageField (Prim . Named . Single $ nameOf (proxy# :: Proxy# a)) Nothing
 
 instance Primitive a => MessageField (UnpackedVec a) where
   encodeMessageField fn = Encode.vectorMessageBuilder (encodePrimitive fn) . unpackedvec
-  decodeMessageField =
-    UnpackedVec . fromList . Foldable.toList <$> repeated decodePrimitive
+
+  decodeMessageField = UnpackedVec . fromList <$> repeated decodePrimitive
+
   protoType _ = messageField (Repeated $ primType (proxy# :: Proxy# a)) (Just DotProto.UnpackedField)
 
 instance forall a. (Named a, Message a) => MessageField (NestedVec a) where
-  encodeMessageField fn = Encode.vectorMessageBuilder (Encode.embedded fn . encodeMessage (fieldNumber 1)) . nestedvec
-  decodeMessageField =
-      fmap (coerce @(Vector a) @(NestedVec a) . fromList . Foldable.toList)
-           (repeated (Decode.embedded' oneMsg))
+  encodeMessageField fn = Encode.vectorMessageBuilder (Encode.embedded fn . encodeMessage 1) . nestedvec
+
+  decodeMessageField = fmap (coerce @(Vector a) @(NestedVec a) . fromList) (repeated (Decode.embedded' oneMsg))
```

Comment:
This is a pre-existing issue, so you can ignore, but:

As in the above code, it might be clearer that coercion is zero-cost if it were outside of `fmap`.